optaplanner-core/src/main/java/org/optaplanner/core/impl/partitionedsearch/DefaultPartitionedSearchPhase.java

```diff
@@ -143,14 +148,44 @@ public void solve(DefaultSolverScope<Solution_> solverScope) {
                 stepEnded(stepScope);
                 phaseScope.setLastCompletedStepScope(stepScope);
             }
-        } finally {
+        } catch (RuntimeException e) {
             // If a partition thread throws an Exception, it is propagated here
             // but the other partition thread won't finish any time soon, so we need to terminate them
             childThreadPlumbingTermination.terminateChildren();
+            throw e;
+        } finally {
+            List<Runnable> partitionSolvers = threadPoolExecutor.shutdownNow();
+            logger.trace("{}         Awaiting termination of {} partition solvers.",
+                    logIndentation, partitionSolvers.size());
+            try {
+                long start = 0;
+                if (logger.isTraceEnabled()) {
+                    start = System.nanoTime();
+                }
+                if (threadPoolExecutor.awaitTermination(1, TimeUnit.MILLISECONDS)) {
+                    if (logger.isTraceEnabled()) {
+                        logger.trace("{}         Thread pool shutdown took {} nanos.",
+                                logIndentation, System.nanoTime() - start);
+                    }
+                } else {
+                    logger.warn("{}Thread pool didn't terminate within the timeout.", logIndentation);
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new IllegalStateException("Thread pool shutdown was interrupted.", e);
```

Comment:
Sorry, this no longer makes sense to me. Here we are on the main thread. I can't see any way how it could be interrupted and THEN start to `awaitTermination()` (while being interrupted). Notice that if a partition thread is interrupted it won't interrupt the solverThread. I don't even have an idea how to test such code. I'll think about it but I suggest this PR be merged without the `isInterrupted()` check.